### PR TITLE
feat(#11151): add byForm qualifier to Report.v1.getUuidsPage

### LINKED ISSUE #11151: Add `byForm` qualifier to `Report.v1` (replaces `medic-client/reports_by_form` page query)
---
title: Add `byForm` qualifier to `Report.v1` (replaces `medic-client/reports_by_form` page query)
type: improvement
priority: medium
domain: data-sync
labels:
  - cht-datasource
  - migration
---

## Description

Replace `db.query('medic-client/reports_by_form', ...)` page-query call sites in cht-core with a typed `byForm(formCode)` qualifier on `Report.v1.getUuidsPage`.

Scope is the page query only. Some callers also use this view with `group: true` for counts, or `key: [form], limit: 1` for sampling — those aggregate variants are out of scope and require a different shaped operation (e.g. `Report.v1.countByForm`). They stay on `db.query` until a separate ticket addresses the aggregate API surface.

Part of the cht-datasource view→qualifier migration series; pattern established in #10973.

## Technical Context

**Components:**
- `shared-libs/cht-datasource/src/qualifier.ts` — add `FormQualifier`
- `shared-libs/cht-datasource/src/report.ts` — extend `getUuidsPage`'s declarative API
- `shared-libs/cht-datasource/src/local/report.ts` — add dispatch arm
- `shared-libs/cht-datasource/src/remote/report.ts` — add dispatch arm
- `api/src/controllers/report.js` — new route `GET /api/v1/report/by-form/:formCode`
- `api/src/routing.js` — wire the route

**Existing References:**
- `shared-libs/cht-datasource/src/local/report.ts` — pattern for the `getUuidsPage` dispatch arms (already has `isFreetextQualifier` arm)
- PR [#10083](https://github.com/medic/cht-core/pull/10083) — added the create/update qualifiers + Local/Remote dispatch. Same pattern as this ticket.

### Adapter Additions

1. **FormQualifier type:**

```typescript
export type FormQualifier = Readonly<{ form: string }>;
export const byForm = (form: string): FormQualifier => {
  if (!isString(form) || form.length === 0) {
    throw new InvalidArgumentError(`Invalid form [${JSON.stringify(form)}].`);
  }
  return { form };
};
export const isFormQualifier = (q: unknown): q is FormQualifier =>
  isRecord(q) && hasField(q, { name: 'form', type: 'string' });
```

The form code is the doc's `form` field (e.g. `pregnancy`, `anc_followup`). Do not normalise.

2. Extend `Report.v1.getUuidsPage` JSDoc for new qualifier shape.

3. Add dispatch arm to `local/report.ts`:
   - When `isFormQualifier(qualifier)`, query `medic-client/reports_by_form` with `key: [qualifier.form]`

4. Add dispatch arm to `remote/report.ts` calling the new endpoint.

5. Add API endpoint and wire it.

6. Expose `byForm` from cht-datasource's `index.ts` under `Qualifier`.

### Caller Sweeps

Audit `grep -rnE "query\(['\"]medic-client/reports_by_form['\"]" webapp/src admin/src shared-libs` before drafting. **Out-of-scope callers** — those passing `group: true` or `reduce: true` or using single-result sampling — stay on `db.query` and are preserved with a comment referencing the follow-up aggregate-API ticket. Only the page-query callers swap to `byForm`.

### Tests

- Unit: `test/qualifier.spec.ts` — `byForm` accepts non-empty strings; rejects empty + non-string
- Unit: `test/local/report.spec.ts` — dispatch arm hits the right view + key
- Unit: `test/remote/report.spec.ts` — endpoint called with correct path
- Caller-side: each modified caller keeps existing tests green
- Pagination: walk two cursor pages with `limit: 5`
- Aggregate-preservation: confirm aggregate callers (if any) are NOT touched by this ticket — `grep -rnE "query\(['\"]medic-client/reports_by_form['\"][^)]*reduce" webapp/src admin/src shared-libs` should return the same matches before and after

## Acceptance Criteria

- `byForm` qualifier in `qualifier.ts` with unit tests passing
- Local and Remote adapters dispatch new qualifier with unit tests passing
- `GET /api/v1/report/by-form/:formCode` endpoint returns `{ data, cursor }`
- Page-query callers swapped; aggregate-query callers explicitly preserved with a comment
- `npm run unit-shared-lib` and `npm run unit-api` pass

## Constraints

- Backwards-compatible: existing calls continue unchanged
- No new top-level `Report.v1` method
- Do not relax `Input.v1.*Input._id?: never`
- Do not normalise the form code
- Aggregate variants (`group: true`, `reduce: true`, `limit: 1` sample) are out of scope. The qualifier-extension pattern doesn't fit count/sample operations — those need narrower operations like `Report.v1.countByForm`. File a follow-up ticket for the aggregate API surface; do not fold it into this ticket.

## References

- PR #10083 — qualifier + dispatch pattern
- PR #9625 — pagination discipline
- Pilot: #10973

### PR DESCRIPTION
**Pull request by ken-talltree-io**
# Description

Adds a `FormQualifier` so report UUIDs can be paged by the forms they were recorded with, replacing direct `medic-client/reports_by_form` page queries.

- `Qualifier.byForms` / `isFormQualifier`, following the shape of `byIds`
- `Report.v1.getUuidsPage` and `getUuids` accept `FreetextQualifier | FormQualifier`
- Local adapter issues a single `keys` query against `medic-client/reports_by_form`; remote adapter sends `?form=` to the existing `GET /api/v1/report/uuid`
- `getUuidsPageByForms` / `getUuidsByForms` on the `Report` `Datasource`

`byForms` is exposed under `Qualifier` automatically — `index.ts` already does `export * as Qualifier from './qualifier'` — so requirement 6 needed no change.

**Revised after review** ([#11327 review](https://github.com/medic/cht-core/pull/11327#issuecomment-5242881009)). Two changes from the first revision, both @jkuester's: the endpoint moved from a `by-form/:formCode` path to a `?form=` query param on the existing route, and the qualifier widened from one form code to many.

**Revised again after [@sugat009's review](https://github.com/medic/cht-core/pull/11327#pullrequestreview-4935755898)**: added the integration coverage this was missing (see the Tests section), fixed a validation gap where a form code with surrounding whitespace silently matched nothing instead of 400ing, closed a dedupe gap for hand-built qualifiers that bypass `byForms()`, and applied the naming/type/comment nitpicks — `FormQualifier`/`isFormQualifier` are now `FormsQualifier`/`isFormsQualifier` to match `IdsQualifier`, `forms` is a non-empty tuple like `contactIds`, and the two invalid-form 400 bodies now share one shape. Full breakdown in [this comment](https://github.com/medic/cht-core/pull/11327#issuecomment-5295130178).

## The caller sweep found nothing in scope

The ticket asks for page-query callers to be swapped. **There are none** — every existing caller of `medic-client/reports_by_form` is an aggregate or a single-result sample, all explicitly out of scope:

| Caller | Shape | Why out of scope |
|---|---|---|
| `api/src/services/impact.js` | `reduce: true, group: true` | aggregate count |
| `api/src/services/export/report-mapper.js:59` | `group: true` | aggregate, enumerates distinct forms |
| `api/src/services/export/report-mapper.js:72` | `key: [form], limit: 1, include_docs: true` | single-result sample |

Each is preserved with a comment pointing at the aggregate API follow-up (#11328), per the ticket.

Three notes on the audit itself — these look templated across the `Datasource View Migration` series rather than specific to this ticket, so the details are in a [comment on #11151](https://github.com/medic/cht-core/issues/11151#issuecomment-5206249403):

1. **The ticket's audit grep searches the wrong directories.** Run verbatim (`... webapp/src admin/src shared-libs`) it returns **zero matches** — every caller is in `api/src`, which isn't in the path. It reports "clean" rather than erroring, so following the ticket literally would mean concluding there are no callers at all.
2. The aggregate-preservation grep matches nothing before **or** after, so it passes trivially while verifying nothing.
3. A fourth reference exists that no `query(...)`-shaped grep can find: `shared-libs/search/src/generate-search-requests.js:66` goes through `getRequestForMultidropdown(...)`. Left untouched — see below, since it drove the multi-form design.

## Design decisions worth a look

**One deliberate deviation from the ticket text.** The `FormQualifier` snippet in the requirements pairs its builder (`!isString(form) || form.length === 0`) with a `hasField` type guard. Those two disagree: `hasField` only checks the field's *type*, so the guard accepts `{ form: '' }` while the builder throws on `''` — and `getUuidsPage`'s assert runs the guard, not the builder. As specified, a blank code returned `200 {"data":[],"cursor":null}` instead of `400`. The guard and the builder now share a single definition of validity, so a blank entry is rejected wherever it enters, including one blank among otherwise valid codes. Internal whitespace is still preserved verbatim, so the "do not normalise" constraint holds.

**Endpoint shape — resolved.** The ticket specified `GET /api/v1/report/by-form/:formCode`, which @jkuester identified as generated rather than intended. `?form=` now sits on the existing `/api/v1/report/uuid` beside `?freetext=`, so every `getUuidsPage` qualifier stays on one route and the route list doesn't grow a segment per qualifier as the series lands. It accepts `?form=a,b` or the parameter repeated, the same as `?ids=` on `/api/v1/report`.

**Multiple form codes, and why the single-form version was the wrong shape.** `generate-search-requests.js:66` already queries this view as `keys: [['P'], ['R']], reduce: false` — multi-key is what the codebase does with `reports_by_form` today. So `FormQualifier` holds an array, deduplicated with order preserved, and the local adapter issues **one `keys` query** rather than one per form.

Order preservation is load-bearing rather than cosmetic: CouchDB groups rows by key in the order supplied, so `skip` is only meaningful across pages while that order is stable. A test pages across a form boundary at a page size that splits it, which is where a per-key rather than whole-result-set skip would surface.

An empty array is rejected rather than returning an empty page — unlike `IdsQualifier`, which accepts one — because it can only mean the caller made a mistake and an empty page would hide it.

**That caller still can't consume the API, and I don't think it can.** `formRequest` builds a view-request *descriptor* that `search.js` executes and then intersects with the date / validity / place filters (`search.js:41,53`). That term needs the complete unpaged id set to intersect against, whereas `getUuidsPage` returns a cursor-paged slice. Migrating it needs either an unpaged variant or the whole filter composition moving behind the datasource — its own ticket, and it touches the offline search path.

**Naming.** The wire parameter is `?form=`; the qualifier field is `forms: string[]`, matching `ids: string[]` on `IdsQualifier`. Each layer is consistent with its neighbours and they differ from each other. Easy to change either side.

**Qualifier precedence.** Freetext is matched first in both adapters, so a qualifier satisfying both shapes behaves exactly as before. Covered by a test in each adapter.

**Error messages preserved.** Widening the assert would have changed the `InvalidArgumentError` message for existing invalid-freetext calls, which the API surfaces in 400 bodies. Only form-shaped input gets the new message; everything else keeps the original `Invalid freetext [...]` text verbatim.

**Unawaited promise.** The form branch returns without awaiting `promisedUseNouveau`, so a binding that only ever served form queries could leave a rejection unobserved. Added an explicit no-op `.catch()`; the freetext branch still awaits and surfaces the real error.

## Tests

- `test/qualifier.spec.ts` — `byForms` dedupes, preserves order, does not normalise, and rejects empty arrays, bare strings, blanks and non-strings
- `test/local/report.spec.ts` — dispatch hits the right view with one `keys` call; freetext machinery untouched
- `test/remote/report.spec.ts` — `?form=` on the shared route, comma joining, cursor omission, freetext precedence
- `test/local/report.spec.ts` — **behavioural** pagination against a key-aware fake view that does not stub page assembly, walked across cursor pages at several limits asserting real UUIDs and cursor values, including **across a form boundary** and with a form that has no reports
- `test/report.spec.ts` — `getDatasource` per-key tests for both new methods, matching the freetext trio
- `api/tests/mocha/controllers/report.spec.js` — comma-separated and repeated `?form=`, single code, empty entries skipped, empty value rejected without querying (now asserting status/message, not just that no query ran), freetext precedence, and the unchanged missing-freetext path
- `test/libs/parameter-validators.spec.ts` — `assertFreetextOrFormsQualifier`, mirroring `assertFreetextQualifier`'s coverage
- `test/local/libs/doc.spec.ts` — `queryDocIdsByKeys` directly, mirroring `queryDocIdsByKey`'s
- `test/local/report.spec.ts` — a hand-built qualifier that bypasses `byForms()` still gets its duplicate form codes dropped before the view query
- `tests/integration/shared-libs/cht-datasource/report.spec.js` and `tests/integration/api/controllers/report.spec.js` — real-CouchDB and HTTP-route coverage for `byForms`/`?form=`, matching the freetext blocks including paging across pages; the HTTP-level invalid-cursor test asserts the 400 the form path throws (it validates synchronously, unlike freetext's 500-from-Nouveau)

Every new behaviour was mutation-tested rather than eyeballed: querying only the first key fails 3 tests, dropping the dedupe fails 1, accepting an empty array fails 2, and sending only the first form remotely fails 1.

That run also caught one of my own tests verifying nothing. `queries a duplicated form only once` passed with the dedupe removed *and* with only the first key queried, because `fetchAndFilterIds` keeps its own `idSet` (`local/libs/doc.ts:184-199`) and drops repeat uuids regardless of what the qualifier does. It was green for a reason unrelated to the change, so I deleted it rather than reword it; the qualifier spec pins the actual dedupe.

## Verification

- `cht-datasource`: **1130 passing, 0 failing**
- `unit-shared-lib` (all shared libs): every lib passes except `shared-libs/outbound`, which misses its 95% nyc threshold (94.85 / 89.55 / 94.89). Byte-identical on `master`, and `outbound` has no dependency on `cht-datasource`. `cht-datasource` meets its own coverage threshold with the new code.
- `unit-api`: **1648 passing, 4 failing**. Ran again at the merge-base (`98162aac3`), rebuilding `cht-datasource` there so the comparison is fair: **1641 passing, 4 failing**, and the failing set is identical — a migration case, `Monitoring service`, and two `Replication Health service` cases. So this change adds 7 passing tests and no failures.
- `shared-libs/search`: **79 passing**, untouched by this change but it shares the view.
- eslint clean on all changed files.

Both acceptance-criteria scripts are green *modulo pre-existing breakage* rather than exiting 0 outright — flagging the distinction rather than claiming a clean run.

**Update:** integration coverage for `?form=` is now added at both `tests/integration/shared-libs/cht-datasource/report.spec.js` and `tests/integration/api/controllers/report.spec.js`, mirroring the freetext blocks. I still don't have a live CHT instance to run them against, so they're unverified — checked by hand against the fixture data and the real `keys`-query semantics the production code already relies on, but not run. Flagging that plainly rather than claiming a green run.

One environment note that cost me time, in case it helps anyone else: `unit-api` only loads if `api/tests/mocha/services/sso-login.spec.js` is passed **first**, exactly as the `package.json` script does. Globbing `api/tests/mocha/**/*.js` on its own dies in `openid-client` with `Cannot use import statement outside a module` before a single test runs.

Closes #11151

# Code review checklist

- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/)
- [x] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/) — the new endpoint is documented inline via its `@openapi` JSDoc block, matching the sibling report routes. Happy to open a cht-docs PR if more is wanted.
- [x] Tested: Unit and/or e2e where appropriate — unit coverage added at the qualifier, local adapter, remote adapter, datasource and controller layers, plus integration coverage at both the shared-libs and HTTP levels (unverified locally; see note above).
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes. — additive only; freetext is matched first in both adapters and the existing `InvalidArgumentError` messages are preserved verbatim.
- [x] AI disclosure: Please disclose use of AI per [the guidelines](https://docs.communityhealthtoolkit.org/community/contributing/ai-guidelines/).

**AI Disclosure:** Written with Claude Code (Claude Opus) — exploring the codebase, drafting the implementation and tests, running the suites and the mutations above. I reviewed the result at each stage and I'm accountable for it.

That review is where most of the corrections came from. In the first revision it caught a validation gap where the builder and type guard disagreed on blank form codes, an unused `assertFormQualifier` export, and missing `getDatasource` per-key tests. In this revision it caught the vacuous test described in the Tests section. Worth stating plainly given the ticket: the `by-form/:formCode` endpoint in the original revision came from following the ticket text literally, and the ticket text turned out to be generated — I flagged the shape as inconsistent at the time but implemented it as written rather than pushing on it, which is the thing I'd do differently.

**This latest round** (addressing @sugat009's review, Claude Sonnet 5) worked through the review comment-by-comment: the integration tests, the two missing unit tests, the two behavioral fixes, and the naming/type/comment nitpicks. For the two behavioral fixes it reverted each one locally and confirmed the corresponding new test actually fails without it, rather than trusting a passing test on its own. I reviewed the diff before it went up and I'm accountable for it; the one thing I couldn't verify myself is the two new integration test files, since I don't have a live CHT instance — said so plainly above rather than claiming a run I don't have.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: shared-libs/cht-datasource/src/local/report.ts [hunk 1]
```diff
@@ -125,21 +133,37 @@ export namespace v1 {
     const queryNouveauFreetext = queryByFreetext(medicDb, 'reports_by_freetext');
     const getOfflineFreetextQueryPageFn = getOfflineFreetextQueryFn(medicDb);
     const promisedUseNouveau = useNouveauIndexes(medicDb);
+    // The form branch below returns without awaiting promisedUseNouveau, so a binding that only ever
+    // serves form queries would leave a rejection unobserved. The freetext branch still awaits (and
+    // so still surfaces) the real error.
+    promisedUseNouveau.catch(() => { /* no-op */ });
+    const queryViewByForms = queryDocIdsByKeys(medicDb, 'medic-client/reports_by_form');
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/local/report.ts:140]: **issue (blocking):** test coverage for `byForms` is below the `freetext` sibling.

`byForms` has no integration test. The `freetext` sibling has integration tests for the same endpoint at two levels:
- Real CouchDB: `tests/integration/shared-libs/cht-datasource/report.spec.js` runs `getUuidsPage` and `getUuids` with `byFreetext`, and it checks paging across pages (lines 198 to 294). `byForms` has none of these.
- HTTP route: `tests/integration/api/controllers/report.spec.js:231` (`GET /api/v1/report/uuid`) checks `freetext` many times. It checks `form` zero times.

Only an integration test runs this real `medic-client/reports_by_form` view. The unit tests use fakes. So they cannot catch a wrong view name, a `key` in place of `keys`, or a change to the view.

Two unit helpers also lack the direct tests that their siblings have:
- `assertFreetextOrFormQualifier` has no test. The sibling `assertFreetextQualifier` has one in `test/libs/parameter-validators.spec.ts`.
- `queryDocIdsByKeys` has no direct test in `test/local/libs/doc.spec.ts`. Both sibling helpers have one. The report spec only stubs it.

The two `form` error tests (`api/tests/mocha/controllers/report.spec.js:244, :272`) check that no query runs, but they do not check the error status or the message. So a change from 400 to 500 would still pass.

Please add integration coverage for `byForms` to match the `freetext` sibling, add the two missing unit tests, and assert the status and message in the error tests.

File: shared-libs/cht-datasource/src/qualifier.ts [hunk 2]
```diff
@@ -176,6 +176,44 @@ export const isKeyedFreetextQualifier = (qualifier: FreetextQualifier): boolean
...
+ * which is never what the caller meant, and silently returning an empty page would hide the mistake.
+ * @param qualifier the qualifier to check
+ * @returns `true` if the given qualifier is a {@link FormQualifier}, otherwise `false`.
+ */
+export const isFormQualifier = (qualifier: unknown): qualifier is FormQualifier => {
+  return isRecord(qualifier)
+    && hasField(qualifier, { name: 'forms', type: 'object' })
+    && Array.isArray(qualifier.forms)
+    && qualifier.forms.length > 0
+    && qualifier.forms.every(form => isString(form) && form.trim().length > 0);
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/qualifier.ts]: **issue (non-blocking):** the validator trims form codes, but the query does not.

`isFormQualifier` accepts a code when `form.trim().length > 0`. `byForms` then stores the code as given, and the view uses the code as given. So `?form=  pregnancy  ` passes validation, but it matches zero reports. I confirmed this on the live API: `?form=pregnancy` returns 2 reports, and the padded form returns an empty page with no error. Is the silent empty page the wanted result? If not, please trim the code before you store it, or reject a code that only passes because of trimming.

File: shared-libs/cht-datasource/src/local/report.ts [hunk 3]
```diff
@@ -125,21 +133,37 @@ export namespace v1 {
...
+        return fetchAndFilterIds(getPageFn, limit)(limit, skip);
       }
 
-      // Use client-side offline freetext views.
+      // The view emits [doc.form], so each form code is a complete key on its own. The qualifier
+      // preserves the caller's order and drops duplicates, which is what keeps `skip` meaningful
+      // from one page to the next — the view returns rows grouped by key in the order supplied.
       const skip = validateCursor(cursor);
-      const getPageFn = getOfflineFreetextQueryPageFn(freetextQualifier);
+      const keys = qualifier.forms.map(form => [form]);
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/local/report.ts]: **note:** the form branch does not remove duplicate form codes.

The `freetext` branch removes duplicates, because it calls `normalizeFreetextQualifier` first. `byForms` removes duplicates too, so the HTTP route and the public `getUuidsByForms` method are safe. But a caller that builds a `FormQualifier` by hand and calls the internal `getUuidsPage` keeps the duplicates. I confirmed the effect on the view: two identical `pregnancy` keys return the rows twice, so the page repeats reports. Removing duplicates here, or in `assertFreetextOrFormQualifier`, closes it.

File: shared-libs/cht-datasource/src/local/report.ts [hunk 4]
```diff
@@ -125,21 +133,37 @@ export namespace v1 {
...
+      const getPageFn = (limit: number, skip: number) => queryViewByForms(keys, limit, skip);
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/local/report.ts:168]: **note (efficiency):** the form page query uses `skip` on `reports_by_form`.

Deep paging is therefore O(skip). There is no forms Nouveau index, so the form path has no bookmark cursor like the freetext server path. This matches the offline freetext branch in the same file, so it is acceptable. It matters only for a form with very many reports paged deeply.

File: api/src/controllers/report.js [hunk 5]
```diff
@@ -19,6 +19,16 @@ const buildIdsQualifier = (ids) => {
   return Qualifier.byIds(idsArray);
 };
 
+// Accepts `?form=a,b` and `?form=a&form=b` alike, matching how `ids` is handled above rather than
+// picking one convention for this parameter alone.
+const buildFormQualifier = (form) => {
+  const formsArray = (Array.isArray(form) ? form : form.split(',')).filter(Boolean);
+  if (!formsArray.length) {
+    throw new InvalidArgumentError(`Invalid forms [${JSON.stringify(form)}].`);
```
Comment (sugat009) [inline api/src/controllers/report.js]: **nitpick:** the 400 body for an invalid form has two shapes.

`?form=` gives `Invalid forms [""]`. `?form=%20%20` gives `Invalid forms [["  "]]`. One is a string, the other is an array. This line throws the first, and `byForms` throws the second. Both are 400.

File: shared-libs/cht-datasource/src/report.ts [hunk 6]
```diff
@@ -406,6 +419,32 @@ export namespace v1 {
...
+      limit?: number | `${number}`
+    ) => Promise<Page<string>>;
+
+    /**
+     * Returns a generator for fetching all the identifiers of reports recorded with any of the given forms.
+     * @param forms the codes of the forms the reports were recorded with (e.g. `['pregnancy']`)
+     * @returns a generator for fetching all matching report identifiers
+     * @throws InvalidArgumentError if the form codes are not a non-empty array of non-blank strings
+     */
+    getUuidsByForms: (forms: string[]) => AsyncGenerator<string, null>;
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/report.ts]: **nitpick:** the `forms` type allows an empty array at compile time.

`getUuidsByForms(forms: string[])` and `FormQualifier` accept an empty array at compile time, but `byForms` rejects it at run time. The sibling `byContactIds([string, ...string[]])` uses a non-empty tuple, so it rejects an empty array at compile time. A non-empty tuple would move the check earlier.

File: shared-libs/cht-datasource/src/qualifier.ts [hunk 7]
```diff
@@ -176,6 +176,44 @@ export const isKeyedFreetextQualifier = (qualifier: FreetextQualifier): boolean
   return false;
 };
 
+/**
+ * A qualifier that identifies entities based on the codes of the forms used to record them.
+ */
+export type FormQualifier = Readonly<{ forms: string[] }>;
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/qualifier.ts]: **nitpick:** the new names mix singular and plural.

The type `FormQualifier` and the guard `isFormQualifier` are singular. The field `forms`, the builder `byForms`, and the methods `getUuidsByForms` are plural. The sibling qualifiers stay plural throughout. The PR and issue title also say "byForm", but the shipped symbol is `byForms`.

File: api/src/services/export/report-mapper.js [hunk 8]
```diff
@@ -55,6 +55,9 @@ module.exports = {
         return Promise.resolve(forms);
       }
 
+      // Aggregate (grouped) query used to enumerate the distinct forms in use. Deliberately left
+      // on db.query by the cht-datasource view->qualifier migration: Qualifier.byForms only covers
+      // the page query — see the aggregate API follow-up ticket.
```
Comment (sugat009) [inline api/src/services/export/report-mapper.js]: **nitpick:** the comments point to an unlinked ticket.

This comment, and the ones at `report-mapper.js:77` and `impact.js:6`, name "the aggregate API follow-up ticket" and `Report.v1.countByForm`, but they give no issue number or link. Please add the issue number.

File: shared-libs/cht-datasource/src/qualifier.ts [hunk 9]
```diff
@@ -176,6 +176,44 @@ export const isKeyedFreetextQualifier = (qualifier: FreetextQualifier): boolean
...
+
+/**
+ * Builds a qualifier for finding entities recorded with any of the given forms.
+ * @param forms the form codes to search with (e.g. `['pregnancy']`). Each is matched verbatim against
+ * the document's `form` field — they are not normalised. Duplicates are removed.
```
Comment (sugat009) [inline shared-libs/cht-datasource/src/qualifier.ts]: **nitpick:** the new comments use em-dashes and British spelling.

This comment and other new comments use em-dashes and the word "normalise". The cht-datasource library uses "normalize" and has no em-dashes in its source. Please match the library style.

File: api/src/controllers/report.js [hunk 10]
```diff
@@ -82,20 +92,33 @@ module.exports = {
...
      *         schema:
      *           type: string
      *           minLength: 3
      *         description: >
      *           A search term for filtering reports. Must be at least 3 characters and not contain whitespace.
+     *           Required unless `form` is given.
+     *       - in: query
+     *         name: form
+     *         required: false
+     *         x-since: 5.3.0
```
Comment (sugat009) [inline api/src/controllers/report.js:117]: **nitpick:** `x-since` on the form parameter is at the parameter level.

Every other `x-since` in the controllers is at the operation level. The placement may be intentional, because the parameter arrived in 5.3.0 while the operation is from 4.18.0. If so, please keep it. If a tool reads only operation-level notes, it may miss this one.

**Review by sugat009 — CHANGES_REQUESTED** [review]
Thank you for this contribution.

This is a good, careful change. It adds a `byForms` qualifier to the report UUID route. It keeps full backward compatibility. The freetext path stays first, and the old error messages do not change. The build passes and the unit tests pass.

**Verified good**
- The build passes. `tsc` reports no errors.
- The unit tests pass. I ran the six changed cht-datasource specs (462 pass) and the api controller spec (19 pass).
- The live API is correct for one form, a comma list, a repeated parameter, the freetext-wins rule, and multi-form paging with no gap and no duplicate.

I request one change before merge: add the integration test coverage that the `freetext` sibling already has (see the blocking comment). Every other comment is non-blocking: one behaviour question, two notes, and several nitpicks.

**Review by sugat009 — APPROVED** [review]
**Re-review: approving. All points addressed.**

Thank you for the thorough fixes. I re-checked each finding against the new head and re-ran the build, tests, and the live API.

- The blocking gap is closed: `byForms` now has integration coverage at both levels, the two missing unit tests are added, and the error tests assert the status and message.
- The behaviour fixes are correct: padded codes are rejected, duplicate forms no longer repeat rows, and both invalid-form bodies match.
- The nitpicks are done, and the follow-up is tracked as `#11328`.

Build clean, 1200 and 19 tests pass, CI green. Nice work.
